### PR TITLE
Closes #5 - Feature/map loader

### LINKED ISSUE #5: rewrite the data loader from txt function
Current load 3d map from txt is really bad need to fix that 

Features required:
- support for loading from different types of file formats 
- support .txt, JSON (if, else .... )

Return:
- boundary: iter6d  
- obstacles: dict(key:iter6d), 
- start: iter3d (default:None)
- goal:iter3d (default:None)

### PR DESCRIPTION
**Pull request by gunjanpatil**
Supports:

- error out if lengths of members don't match
- error out if inavlid string are present
- what to return if start and obstacles are not given?
- raise error if boundary, goal is not given
- warn for multiple obstacles of same values

### CODE REVIEW

File: pybotic/utils/world_utils.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    given path to file, load 3D world map
+
+    Args:
+        file_name (str): Path to 3D world map data
+
+    Returns:
+        boundary (numpy.ndarray, shape=(6,)): physical limits of the world
+        obstacles (numpy.ndarray, shape=(6,)): physical bounds of obstacles
+        start (numpy.ndarray, shape=3,)): start location
+        goal (numpy.ndarray, shape=(3,)): goal location
```
Comment (00arun00) [inline pybotic/utils/world_utils.py:19]: good comments

File: pybotic/utils/world_utils.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    # to error out for repeating arguments
+    key_words = {}
+    # Set of obstacles to warn for repeating obstacle values
+    obstacles = set()
+    # Read all lines from file
+    with open(file_name) as f:
+        for i, line in enumerate(f.readlines()):
+            sl = re.split(': |, |\n| |,', line)[:-1]
+            word = sl[0]
+            if word == "#" or not word:
```
Comment (00arun00) [inline pybotic/utils/world_utils.py:67]: add comment 
skipping comments or empty lines

File: pybotic/utils/world_utils.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+                continue
+            array = str_conversion(sl[1:])
+            # taking care of invalid statements
+            if len(array) == 0:
+                raise SyntaxError("Invalid statement \"{}\"".format(word))
+            # taking care of boundary, start, goal and their repetitions
+            if word in ["boundary", "start", "goal"]:
+                if word not in key_words:
```
Comment (00arun00) [inline pybotic/utils/world_utils.py:75]: add a comment why you are doing it

File: pybotic/utils/world_utils.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+                    key_words[word] = str_conversion(array)
+                else:
+                    raise ValueError(f"File has mutiple {word} keyword")
+            # taking care of obstacles
+            elif word == "obstacle":
+                obstacle = tuple(array)
+                key = "obstacles"
+                if obstacle in obstacles:
```
Comment (00arun00) [inline pybotic/utils/world_utils.py:84]: comment

File: pybotic/utils/world_utils.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+                    warnings.warn(f"Repeating obstacle {array}, if not"
+                                  f" expected, remove them", RuntimeWarning)
+                if key not in key_words:
```
Comment (00arun00) [inline pybotic/utils/world_utils.py:89]: comment

File: pybotic/utils/world_utils.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+                    key_words[key] = {word+'_0': array}
+                elif key in key_words:
```
Comment (00arun00) [inline pybotic/utils/world_utils.py:93]: comment

File: pybotic/utils/world_utils.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+                    key_words[key][word+'_'+str(len(key_words[key]))] = array
+                obstacles.add(tuple(array))
+            # taking care of invalid key words
+            else:
+                raise ValueError("Invalid key word, not in "
```
Comment (00arun00) [inline pybotic/utils/world_utils.py]: display the {key}

File: pybotic/utils/__init__.py [hunk 8]
```diff
@@ -0,0 +1 @@
+from . import world_utils
```
Comment (00arun00) [inline pybotic/utils/__init__.py]: update a to remove access to load_txt and other funcitons

File: pybotic/utils/world_utils.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+    Raises:
+        ValueError: if boundary, goal or both not found
+        ValueError: if required lengths do not match
+
+    Warnings:
+        RuntimeWarning: if start location is not given
+    """
+    if 'boundary' not in key_words and 'goal' not in key_words:
+        raise ValueError("boundary and goal not specified in the file")
```
Comment (00arun00) [inline pybotic/utils/world_utils.py]: no need

File: pybotic/utils/world_utils.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    if 'boundary' not in key_words:
+        raise ValueError("boundary not specified in the file")
+    if 'goal' not in key_words:
+        raise ValueError("goal not specified in the file")
+
+    for key, val in key_words.items():
+        if key == 'boundary':
+            if len(val) != 6:
+                raise ValueError(f"Invalid {key} value, has {len(val)} items")
```
Comment (00arun00) [inline pybotic/utils/world_utils.py]: give expected also

File: pybotic/utils/world_utils.py [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+        if key == 'start' or key == 'goal':
+            if len(val) != 3:
+                raise ValueError(f"Invalid {key} value, has {len(val)} items")
```
Comment (00arun00) [inline pybotic/utils/world_utils.py]: give expect also

File: pybotic/utils/world_utils.py [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+    if "start" not in key_words:
+        warnings.warn("start loc not given,assuming (0, 0, 0)", RuntimeWarning)
```
Comment (00arun00) [inline pybotic/utils/world_utils.py]: userwarming

File: pybotic/utils/world_utils.py [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+        key_words["start"] = np.zeros((3))
+
+    if "obstacles" not in key_words:
+        key_words["obstacles"] = None
+    else:
+        for i, (obs, val) in enumerate(key_words["obstacles"].items()):
+            if len(val) != 6:
+                raise ValueError(f"Invalid obstacle value,"
+                                 f" has {len(val)} items")
```
Comment (00arun00) [inline pybotic/utils/world_utils.py]: expected

File: pybotic/utils/world_utils.py [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+    Args:
+        arr (list[str]): List of str to be converted
+
+    Returns:
+        converted numpy.ndarray[float]
+
+    Raises:
+        ValueError: if elements like str/empty are not convertible to float
+    """
+    return np.array(arr).astype(np.float)
```
Comment (00arun00) [inline pybotic/utils/world_utils.py:174]: shouldn't be needed but ok

File: pybotic/utils/world_utils.py [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+    Given a list of strings, convert it to numpy array of float
+
+    Args:
+        arr (list[str]): List of str to be converted
+
+    Returns:
+        converted numpy.ndarray[float]
+
+    Raises:
+        ValueError: if elements like str/empty are not convertible to float
```
Comment (00arun00) [inline pybotic/utils/world_utils.py:172]: need not

File: tests/sample_world.txt [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+obstacle: 0.0, 2.0, 4.5, 10.0, 2.5, 6.0
+
+obstacle: 7.0, 2.0, 1.5, 10.0, 2.5, 5.0
+
+obstacle: 6.0, 1.0, -0.5, 2.0, -1.2, 0.5
+
+obstacle: 5.0, 1.0, -0.5, 2.0, -1.2, 0.5
+
+start: 0.0, 0.0, 0.0
+goal: 5.0, -3.0, 2.0
```
Comment (00arun00) [inline tests/sample_world.txt]: frozen for now, also add to issue chat

**Review by 00arun00 — CHANGES_REQUESTED** [review]
mostly add comments

File: pybotic/utils/world_utils.py [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+
+    Warnings:
+        RuntimeWarning: if start location is not given
+    """
+    if 'boundary' not in key_words and 'goal' not in key_words:
+        raise ValueError("boundary and goal not specified in the file")
+    if 'boundary' not in key_words:
+        raise ValueError("boundary not specified in the file")
+    if 'goal' not in key_words:
+        raise ValueError("goal not specified in the file")
```
Comment (00arun00) [inline pybotic/utils/world_utils.py]: no need

File: pybotic/utils/world_utils.py (hunk 2 above)
Comment (gunjanpatil) [reply pybotic/utils/world_utils.py:67]: done

**Review by 00arun00 — APPROVED** [review]
Yaaayyy :)

**Pull request merged**